In [68]:
# importing libraries
import pandas as pd
import numpy as np

In [69]:
# loading dataset
pdr_locations = pd.read_csv(r'C:\Users\Камила\Big_data\charging_stations\dataset\full\pdr_locations.csv')

In [70]:
# look at first 3 rows
pdr_locations.head(3)

connector_id  connector_station_id     connector_evse_id  \
0        462849                385079  IT*F2X*EF2XITA2151*1   
1        462850                385079  IT*F2X*EF2XITA2151*2   
2        462851                385079  IT*F2X*EF2XITA2151*3   

   connector_connector_status_id connector_status  \
0                              1        AVAILABLE   
1                              1        AVAILABLE   
2                              1        AVAILABLE   

                          connector_uid  station_id  \
0  6ea4dcd0-6678-4585-8b8d-41ea3a1cee39      385079   
1  826d0eaa-0de0-4aa4-b65e-5db33380c57a      385079   
2  bdb4f814-1754-402e-91fc-ab020f180e1a      385079   

                            station_uid                   station_name  \
0  fe01af77-ad39-4ad9-80bf-322d7c02a4a9  Area di Servizio - Arda Ovest   
1  fe01af77-ad39-4ad9-80bf-322d7c02a4a9  Area di Servizio - Arda Ovest   
2  fe01af77-ad39-4ad9-80bf-322d7c02a4a9  Area di Servizio - Arda Ovest   

            station_address  ...  totSessionsNotLocal  totSessionsLocal  \
0  A1 - Autostrada del Sole  ...                  198                 0   
1  A1 - Autostrada del Sole  ...                   79                 0   
2  A1 - Autostrada del Sole  ...                  289                 0   

   totCost totHouseSessions totPublicSessions  totPublicCost  totHouseCost  \
0  2312.77                0               201        2312.77             0   
1   700.40                0                80         700.40             0   
2   944.33                0               297         944.33             0   

   totHouseEnergy  totPublicEnergy connector_isSuspended  
0               0         4195.081                   NaN  
1               0         1273.441                   NaN  
2               0         2097.308                   NaN  

[3 rows x 144 columns]

### Inspecting dataset

In [71]:
# number of rows, columns
pdr_locations.shape

(830, 144)

In [72]:
# look at the general statistics
pdr_locations.describe()

connector_id  connector_station_id  connector_connector_status_id  \
count  8.300000e+02          8.300000e+02                     830.000000   
mean   9.240990e+05          6.863694e+05                       1.151807   
std    3.095176e+05          2.529324e+05                       0.873737   
min    4.187280e+05          3.691160e+05                       1.000000   
25%    6.250482e+05          4.446590e+05                       1.000000   
50%    9.743725e+05          5.807380e+05                       1.000000   
75%    1.167045e+06          9.524130e+05                       1.000000   
max    1.413007e+06          1.033694e+06                       9.000000   

         station_id  station_street_number  station_postal_code  \
count  8.300000e+02                    0.0           830.000000   
mean   6.863694e+05                    NaN         36763.586747   
std    2.529324e+05                    NaN         24449.650843   
min    3.691160e+05                    NaN            10.000000   
25%    4.446590e+05                    NaN         20026.000000   
50%    5.807380e+05                    NaN         33010.000000   
75%    9.524130e+05                    NaN         52041.000000   
max    1.033694e+06                    NaN         83100.000000   

       station_coordinates_lat  station_coordinates_lon  station_directions  \
count               830.000000               830.000000                 0.0   
mean                 43.772645                11.679412                 NaN   
std                   1.605664                 2.172102                 NaN   
min                  40.837613                 8.440903                 NaN   
25%                  42.200472                 9.393760                 NaN   
50%                  44.392512                11.570791                 NaN   
75%                  45.074900                13.334975                 NaN   
max                  46.499013                16.769807                 NaN   

       station_time_zone  ...  totSessions  totSessionsNotLocal  \
count                0.0  ...   830.000000           830.000000   
mean                 NaN  ...   465.036145           454.827711   
std                  NaN  ...   490.645257           480.294118   
min                  NaN  ...     0.000000             0.000000   
25%                  NaN  ...    79.000000            77.000000   
50%                  NaN  ...   283.500000           276.500000   
75%                  NaN  ...   729.750000           715.750000   
max                  NaN  ...  3062.000000          2981.000000   

       totSessionsLocal       totCost  totHouseSessions  totPublicSessions  \
count        830.000000    830.000000             830.0         830.000000   
mean           0.002410   9467.685855               0.0         465.036145   
std            0.049058  10802.687982               0.0         490.645257   
min            0.000000      0.000000               0.0           0.000000   
25%            0.000000    584.035000               0.0          79.000000   
50%            0.000000   5360.650000               0.0         283.500000   
75%            0.000000  15833.222500               0.0         729.750000   
max            1.000000  64974.670000               0.0        3062.000000   

       totPublicCost  totHouseCost  totHouseEnergy  totPublicEnergy  
count     830.000000         830.0           830.0       830.000000  
mean     9467.685855           0.0             0.0     14006.847572  
std     10802.687982           0.0             0.0     15775.476469  
min         0.000000           0.0             0.0         0.000000  
25%       584.035000           0.0             0.0      1134.687000  
50%      5360.650000           0.0             0.0      7879.400000  
75%     15833.222500           0.0             0.0     23273.082000  
max     64974.670000           0.0             0.0     95445.220000  

[8 rows x 85 columns]

### Handling missing values

In [73]:
# check for columns with missing values
missing_data = pdr_locations.isnull().sum().sort_values(ascending=False)
missing_data

station_owner_company_parent_company_id                 830
station_owner_company_user_id                           830
station_owner_company_deadline                          830
station_owner_company_id_offer                          830
station_owner_company_pin_color                         830
station_owner_company_endpoint                          830
station_owner_company_created_at                        830
station_owner_company_heycharge_email                   830
station_owner_company_heycharge_api_key                 830
station_payment_terminal_sn                             830
station_mid_meter_sn                                    830
station_owner_company_promotion_date                    830
station_owner_company_stars                             830
station_owner_company_vat                               830
station_owner_company_stripe_invoice_extra_fields       830
station_owner_company_stripe_public_key                 830
station_owner_company_stripe_customer_id

In [74]:
# check duplications
pdr_locations.duplicated().sum()

0

we know that there are 830 rows overall in the dataset, and I will drop the columns that have all these rows empty.

In [75]:
# first, i want to count the number of these empty columns
num_empty_columns = len(pdr_locations.columns[pdr_locations.isnull().sum() == len(pdr_locations)])
num_empty_columns

50

In [76]:
# from above i know that there are 50 empty columns, i will drop them. 
columns_to_drop = pdr_locations.columns[pdr_locations.isnull().sum() == len(pdr_locations)]

In [77]:
pdr_locations = pdr_locations.drop(columns=columns_to_drop)
pdr_locations.shape

(830, 94)

In [78]:
# check for columns with missing values
missing_data = pdr_locations.isnull().sum().sort_values(ascending=False)
missing_data

connector_isSuspended                                819
station_commissioning_date                           545
station_installation_date                            122
station_serial_number                                 79
station_firmware                                      79
station_model                                         79
station_brand                                         79
station_type_status                                    2
station_owner_company_owner_cost_per_kwh               0
station_owner_company_brand_color                      0
station_owner_company_show_map                         0
station_owner_company_show_roaming                     0
station_owner_company_show_advanced_services           0
station_owner_company_emp_services                     0
station_owner_company_is_dealer                        0
station_owner_company_is_lifestyle                     0
station_owner_company_is_evfleet                       0
station_owner_company_is_subcpo

as there are 830 rows, 50% of them will be 415. so i will drop columns that have more than 50% empty rows.

In [79]:
pdr_locations = pdr_locations.dropna(axis=1, thresh=len(pdr_locations) * 0.5)

In [80]:
pdr_locations.shape

(830, 92)

In [81]:
# displaying more rows of missing data
pd.set_option('display.max_rows', None)

# check for columns with missing values
missing_data = pdr_locations.isnull().sum().sort_values(ascending=False)
missing_data

station_installation_date                            122
station_serial_number                                 79
station_firmware                                      79
station_model                                         79
station_brand                                         79
station_type_status                                    2
connector_id                                           0
station_owner_company_is_lifestyle                     0
station_owner_company_show_map                         0
station_owner_company_show_roaming                     0
station_owner_company_show_advanced_services           0
station_owner_company_emp_services                     0
station_owner_company_is_dealer                        0
station_owner_company_is_evfleet                       0
station_owner_company_pre_authorization_amount         0
station_owner_company_is_subcpo                        0
station_owner_company_show_emp_cost                    0
station_owner_company_blacklist

In [82]:
# i'm gonna focus only on missing valued columns
missing_columns = missing_data[missing_data > 0]
missing_columns

station_installation_date    122
station_serial_number         79
station_firmware              79
station_model                 79
station_brand                 79
station_type_status            2
dtype: int64

looking at the columns that have 70 missing rows, i want to check if they are consistent across these 4 columns:

In [83]:
# identifying 4 columns with 79 missing values
columns_with_79_missing = ['station_serial_number', 'station_firmware', 'station_model', 'station_brand']

# check if the missing values are consistent across these columns
missing_in_consistent_rows = pdr_locations[columns_with_79_missing].isnull().all(axis=1)

# filter rows where all 4 columns have missing values
consistent_missing_rows = pdr_locations[missing_in_consistent_rows]
consistent_missing_rows


connector_id  connector_station_id       connector_evse_id  \
227       1354231               1014238    IT*F2X*EF2XITN0503*1   
240       1354243               1014240    IT*F2X*EF2XITN0603*1   
241       1354244               1014240    IT*F2X*EF2XITN0603*2   
242       1354245               1014241    IT*F2X*EF2XITN0604*1   
243       1354246               1014241    IT*F2X*EF2XITN0604*2   
261       1402425               1030173    IT*F2X*EF2XITA7103*1   
267       1402426               1030173    IT*F2X*EF2XITA7103*2   
318       1352425               1013302    IT*F2X*EF2XITD2104*1   
319       1352426               1013302    IT*F2X*EF2XITD2104*2   
322       1352435               1013309    IT*F2X*EF2XITI1104*1   
323       1352436               1013309    IT*F2X*EF2XITI1104*2   
324       1352437               1013310    IT*F2X*EF2XITI1103*1   
325       1352438               1013310    IT*F2X*EF2XITI1103*2   
352       1352497               1013332  IT*F2X*EF2XITCOLMI01*1   
353       1352498               1013332  IT*F2X*EF2XITCOLMI01*2   
354       1352499               1013333  IT*F2X*EF2XITCOLMI02*1   
358       1352500               1013333  IT*F2X*EF2XITCOLMI02*2   
359       1352501               1013334  IT*F2X*EF2XITCOLMI03*1   
372       1352508               1013337  IT*F2X*EF2XITCOLMI06*2   
383       1184592                957773    IT*F2X*EF2XITD3001*1   
387       1184593                957773    IT*F2X*EF2XITD3001*2   
388       1184594                957774    IT*F2X*EF2XITD3002*1   
389       1184595                957774    IT*F2X*EF2XITD3002*2   
390       1184596                957775    IT*F2X*EF2XITD3051*1   
391       1184597                957775    IT*F2X*EF2XITD3051*2   
392       1184598                957775    IT*F2X*EF2XITD3051*3   
437       1389890               1024988    IT*F2X*EF2XITH3004*1   
439       1389891               1024988    IT*F2X*EF2XITH3004*2   
483       1362240               1017302    IT*F2X*EF2XITA3551*1   
484       1362241               1017302    IT*F2X*EF2XITA3551*2   
485       1362242               1017302    IT*F2X*EF2XITA3551*3   
521       1352502               1013334  IT*F2X*EF2XITCOLMI03*2   
522       1352503               1013335  IT*F2X*EF2XITCOLMI04*1   
523       1352504               1013335  IT*F2X*EF2XITCOLMI04*2   
524       1352505               1013336  IT*F2X*EF2XITCOLMI05*1   
525       1352506               1013336  IT*F2X*EF2XITCOLMI05*2   
526       1389888               1024987    IT*F2X*EF2XITH3003*1   
530       1352507               1013337  IT*F2X*EF2XITCOLMI06*1   
531       1389889               1024987    IT*F2X*EF2XITH3003*2   
547       1215844                967953    IT*F2X*EF2XITN2501*1   
548       1215845                967954    IT*F2X*EF2XITN2502*1   
549       1215846                967955    IT*F2X*EF2XITN2503*1   
550       1215847                967956    IT*F2X*EF2XITN2504*1   
562       1144233                945035    IT*F2X*EF2XITD1101*1   
563       1144234                945035    IT*F2X*EF2XITD1101*2   
564       1144235                945036    IT*F2X*EF2XITD1102*1   
565       1144236                945036    IT*F2X*EF2XITD1102*2   
566       1144237                945037    IT*F2X*EF2XITD1151*1   
567       1144238                945037    IT*F2X*EF2XITD1151*2   
568       1144239                945037    IT*F2X*EF2XITD1151*3   
642       1402427               1030174    IT*F2X*EF2XITA7104*1   
643       1402428               1030174    IT*F2X*EF2XITA7104*2   
644       1402429               1030175    IT*F2X*EF2XITA7003*1   
645       1402430               1030175    IT*F2X*EF2XITA7003*2   
650       1224268                970936    IT*F2X*EF2XITA2502*1   
653       1224269                970936    IT*F2X*EF2XITA2502*2   
657       1224279                970941    IT*F2X*EF2XITA2601*1   
658       1224280                970941    IT*F2X*EF2XITA2601*2   
659       1224281                970942    IT*F2X*EF

above i show exactly those 79 rows, but here i'm making sure if it's true.

In [84]:
# identifying the 4 columns with 79 missing values
columns_with_79_missing = ['station_serial_number', 'station_firmware', 'station_model', 'station_brand']

# check if the missing values are consistent across these columns
consistent_missing = pdr_locations[columns_with_79_missing].isnull().all(axis=1).sum() == 79
consistent_missing

True

so, the result is yes, but still i want to see if they directly affect on customer behaviour, that is why i will see their unique values, and if they have too many, i would like to drop them, as it will be hard to analyze.

In [85]:
# check the number of unique values in each column
for column in ['station_brand', 'station_model', 'station_firmware', 'station_serial_number']:
    unique_values = pdr_locations[column].nunique()
    print(f"{column}: {unique_values} unique values")


station_brand: 4 unique values
station_model: 11 unique values
station_firmware: 14 unique values
station_serial_number: 333 unique values


+ thinking logically, it is obvious that *station_brand* could be **important** for the customer, especially if customers prefer certain brands.
+ *station_model* could be important if specific station models offer better features, but it's **less likely** to directly influence customer behavior unless advertised as such;
+ *station_firmware* usually pertains to the software version of the station, which is **unlikely** to directly impact customer decisions, unless customers are aware that firmware impacts performance.
+ *station_serial_number* is the serial number for charging station hardware and **would not directly** affect customer behavior.

That's why I can leave station_brand, and drop other 3 as they either don't significantly influence customer behavior or have too many unique values to be meaningful.

In [86]:
# dropping the columns that i don't need
columns_to_drop = ['station_model', 'station_firmware', 'station_serial_number']

# drop these columns from the dataset
pdr_locations = pdr_locations.drop(columns=columns_to_drop)
pdr_locations.head()


connector_id  connector_station_id     connector_evse_id  \
0        462849                385079  IT*F2X*EF2XITA2151*1   
1        462850                385079  IT*F2X*EF2XITA2151*2   
2        462851                385079  IT*F2X*EF2XITA2151*3   
3        966351                578546  IT*F2X*EF2XITN3001*1   
4        966352                578546  IT*F2X*EF2XITN3001*2   

   connector_connector_status_id connector_status  \
0                              1        AVAILABLE   
1                              1        AVAILABLE   
2                              1        AVAILABLE   
3                              1        AVAILABLE   
4                              1        AVAILABLE   

                          connector_uid  station_id  \
0  6ea4dcd0-6678-4585-8b8d-41ea3a1cee39      385079   
1  826d0eaa-0de0-4aa4-b65e-5db33380c57a      385079   
2  bdb4f814-1754-402e-91fc-ab020f180e1a      385079   
3  bdf9db5b-92be-4007-b729-0dcc1d73196e      578546   
4  f03b3c38-b0ca-4aed-8f9d-f03c133f7e19      578546   

                            station_uid                     station_name  \
0  fe01af77-ad39-4ad9-80bf-322d7c02a4a9    Area di Servizio - Arda Ovest   
1  fe01af77-ad39-4ad9-80bf-322d7c02a4a9    Area di Servizio - Arda Ovest   
2  fe01af77-ad39-4ad9-80bf-322d7c02a4a9    Area di Servizio - Arda Ovest   
3  addf1a35-1966-494a-8eea-8d9448eb8013  Area di Servizio - Rubicone Est   
4  addf1a35-1966-494a-8eea-8d9448eb8013  Area di Servizio - Rubicone Est   

              station_address  ...  totSessions totSessionsNotLocal  \
0    A1 - Autostrada del Sole  ...          201                 198   
1    A1 - Autostrada del Sole  ...           80                  79   
2    A1 - Autostrada del Sole  ...          297                 289   
3  A14 - Autostrada Adriatica  ...         1220                1199   
4  A14 - Autostrada Adriatica  ...         1096                1072   

  totSessionsLocal   totCost  totHouseSessions  totPublicSessions  \
0                0   2312.77                 0                201   
1                0    700.40                 0                 80   
2                0    944.33                 0                297   
3                0  27640.46                 0               1220   
4                0  25102.30                 0               1096   

  totPublicCost  totHouseCost totHouseEnergy totPublicEnergy  
0       2312.77             0              0        4195.081  
1        700.40             0              0        1273.441  
2        944.33             0              0        2097.308  
3      27640.46             0              0       40645.668  
4      25102.30             0              0       36914.252  

[5 rows x 89 columns]

In [87]:
# displaying more rows of missing data
pd.set_option('display.max_rows', None)

# check for columns with missing values
missing_data = pdr_locations.isnull().sum().sort_values(ascending=False)

# i'm gonna focus only on missing valued columns
missing_columns = missing_data[missing_data > 0]
missing_columns

station_installation_date    122
station_brand                 79
station_type_status            2
dtype: int64

now i want to see what kind of unique values does station_brand have:

In [88]:
# display the unique values of the 'station_brand' column
unique_brands = pdr_locations['station_brand'].unique()
unique_brands

array(['EVTEC', 'alpitronic GmbH', nan, 'FIMER SPA', 'Wall Box Chargers'],
      dtype=object)

In [89]:
# replace nan values in 'station_brand' with 'Unknown'
pdr_locations['station_brand'].fillna('Unknown', inplace=True)

# verify the replacement by checking the unique values again
unique_brands_after_replacement = pdr_locations['station_brand'].unique()
print(unique_brands_after_replacement)


['EVTEC' 'alpitronic GmbH' 'Unknown' 'FIMER SPA' 'Wall Box Chargers']


In [90]:
# display the unique values of the 'station_type_status' column
unique_brands = pdr_locations['station_type_status'].unique()
unique_brands

array(['public', 'to_be_installed', nan], dtype=object)

there are 2 rows that have empty station status, and i cannot fill them with either 'public' or 'to be installed', that's why i decided to drop them, it will not significantly affect the results.

In [91]:
# Drop rows where 'station_type_status' is NaN
pdr_locations.dropna(subset=['station_type_status'], inplace=True)

however, about station_installation_date I am not sure, there are 122 missing values, maybe we will need this information further, so i just want to fill it for now with '0000-00-00T00:00:00.000000Z'

In [92]:
# replace NaN values in 'station_installation_date' with the placeholder date
pdr_locations['station_installation_date'].fillna('0000-00-00T00:00:00.000000Z', inplace=True)

# verify the replacement by checking unique values in 'station_installation_date'
unique_installation_dates = pdr_locations['station_installation_date'].unique()

In [93]:
# displaying more rows of missing data
pd.set_option('display.max_rows', None)

# check for columns with missing values
missing_data = pdr_locations.isnull().sum().sort_values(ascending=False)

# i'm gonna focus only on missing valued columns
missing_columns = missing_data[missing_data > 0]
missing_columns

Series([], dtype: int64)

In [94]:
pdr_locations.shape

(828, 89)

now, i want to see other columns' unique values, because if there is only one unique value, i don't think it's a good idea to leave this column in dataset. especially, if it's connected like with Italy, because we all work on the dataset of one country, and we want our dataset to have more meaningful data.

In [95]:
# calculate the number of unique values in each column
unique_values_count = pdr_locations.nunique()

# convert to a DataFrame for better visualization
unique_values_df = pd.DataFrame(unique_values_count, columns=['Number of Unique Values'])
unique_values_df


Number of Unique Values
connector_id                                                           828
connector_station_id                                                   370
connector_evse_id                                                      828
connector_connector_status_id                                            4
connector_status                                                         4
connector_uid                                                          828
station_id                                                             370
station_uid                                                            370
station_name                                                           112
station_address                                                         38
station_postal_code                                                     68
station_city                                                            67
station_country_id                                                       1
station_country                                                          1
station_coordinates_lat                                                347
station_coordinates_lon                                                350
station_type_name                                                        1
station_station_type_id                                                  1
station_brand                                                            5
station_type_status                                                      2
station_installation_date                                              120
station_is_offline                                                       1
station_is_at_home                                                       1
station_is_qr_private_enable                                             1
station_owner_company_id                                                 1
station_is_visible                                                       2
station_owner_company_uid                                                1
station_owner_company_name                                               1
station_owner_company_eMI3_id                                            1
station_owner_company_imported                                           1
station_owner_company_description_en                                     1
station_owner_company_description_it                                     1
station_owner_company_description_de                                     1
station_owner_company_description_fr                                     1
station_owner_company_street                                             1
station_owner_company_street_number                                      1
station_owner_company_zip                                                1
station_owner_company_city                                               1
station_owner_company_district                                           1
station_owner_company_state                                              1
station_owner_company_country                                            1
station_owner_company_company_types_id                                   1
station_owner_company_primary_category_id                                1
station_owner_company_stripe_onboarding_status_id                        1
station_owner_company_is_always_open                                     1
station_owner_company_has_contract                                       1
station_owner_company_is_promotion                                       1
station_owner_company_has_colored_marker                                 1
station_owner_company_corporate_admin_id                                 1
station_owner_company_logo                                               1
station_owner_company_updated_at                                         1
station_owner_company_is_jlr                                             1
station_owner_company_is_shopping_center                                 1
s

we see there are a lot of columns that have only one unique value, now i will show what are they:

In [96]:
# filter columns that have only 1 unique value
single_value_columns = pdr_locations.nunique()[pdr_locations.nunique() == 1].index

# create a dataFrame to show the column name, number of unique values, and the value itself
single_value_df = pd.DataFrame({
    'Column Name': single_value_columns,
    'Number of Unique Values': [1] * len(single_value_columns),
    'Value': [pdr_locations[col].unique()[0] for col in single_value_columns]
}).reset_index(drop=True)

print(single_value_df.to_string(index=False))

                                      Column Name  Number of Unique Values                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    Value
                               station_country_id                        1                                                                                                                                                                                                                                                                                                                                                                          

In [97]:
# Drop the columns that have only one unique value from the CSV DataFrame
pdr_locations = pdr_locations.drop(columns=single_value_columns)

# Display the remaining columns in the cleaned DataFrame
pdr_locations.columns

Index(['connector_id', 'connector_station_id', 'connector_evse_id',
       'connector_connector_status_id', 'connector_status', 'connector_uid',
       'station_id', 'station_uid', 'station_name', 'station_address',
       'station_postal_code', 'station_city', 'station_coordinates_lat',
       'station_coordinates_lon', 'station_brand', 'station_type_status',
       'station_installation_date', 'station_is_visible', 'plugs', 'sessions',
       'totEnergy', 'totEnergyLocal', 'totEnergyNotLocal', 'totSessions',
       'totSessionsNotLocal', 'totSessionsLocal', 'totCost',
       'totPublicSessions', 'totPublicCost', 'totPublicEnergy'],
      dtype='object')

In [98]:
pdr_locations.shape

(828, 30)

In [101]:
pdr_locations.head(3)

connector_id  connector_station_id     connector_evse_id  \
0        462849                385079  IT*F2X*EF2XITA2151*1   
1        462850                385079  IT*F2X*EF2XITA2151*2   
2        462851                385079  IT*F2X*EF2XITA2151*3   

   connector_connector_status_id connector_status  \
0                              1        AVAILABLE   
1                              1        AVAILABLE   
2                              1        AVAILABLE   

                          connector_uid  station_id  \
0  6ea4dcd0-6678-4585-8b8d-41ea3a1cee39      385079   
1  826d0eaa-0de0-4aa4-b65e-5db33380c57a      385079   
2  bdb4f814-1754-402e-91fc-ab020f180e1a      385079   

                            station_uid                   station_name  \
0  fe01af77-ad39-4ad9-80bf-322d7c02a4a9  Area di Servizio - Arda Ovest   
1  fe01af77-ad39-4ad9-80bf-322d7c02a4a9  Area di Servizio - Arda Ovest   
2  fe01af77-ad39-4ad9-80bf-322d7c02a4a9  Area di Servizio - Arda Ovest   

            station_address  ...  totEnergy totEnergyLocal  totEnergyNotLocal  \
0  A1 - Autostrada del Sole  ...   4195.081            0.0           4064.325   
1  A1 - Autostrada del Sole  ...   1273.441            0.0           1256.926   
2  A1 - Autostrada del Sole  ...   2097.308            0.0           2064.748   

   totSessions totSessionsNotLocal totSessionsLocal  totCost  \
0          201                 198                0  2312.77   
1           80                  79                0   700.40   
2          297                 289                0   944.33   

   totPublicSessions totPublicCost totPublicEnergy  
0                201       2312.77        4195.081  
1                 80        700.40        1273.441  
2                297        944.33        2097.308  

[3 rows x 30 columns]

For the cleaning it's done, but it need to be rechecked for encoding and for these 1 unique valued columns.

### Visualization part